pkg/elle/txn/helper_fn.go

```diff
@@ -0,0 +1,150 @@
+package txn
+
+import (
+	"hash/fnv"
+	"sort"
+	"strconv"
+
+	"github.com/pingcap/tipocket/pkg/elle/core"
+)
+
+type OpMopIterator struct {
+	history      core.History
+	historyIndex int
+	mopIndex     int
+}
+
+func (omi *OpMopIterator) Next() (*core.Op, core.Mop) {
+	op, mop := &omi.history[omi.historyIndex], omi.history[omi.historyIndex].Value[omi.mopIndex]
+
+	// inc index
+	if omi.mopIndex == len(omi.history[omi.historyIndex].Value) {
+		omi.mopIndex = 0
+		omi.historyIndex++
+	} else {
+		omi.mopIndex++
+	}
+	return op, mop
+}
+
+func (omi *OpMopIterator) HasNext() bool {
+	return omi.historyIndex < len(omi.history)
```

Comment:
```suggestion
	return omi.historyIndex < len(omi.history) || omi.mopIndex < len(omi.history[omi.historyIndex].Value)
```

Reply:
Seems the condition isn't completely right. @mapleFU

Reply:
After the fix of Next(), it's complete right now.

Reply:
I think it's right, when the iterate is really finished, `omi.historyIndex` should equal to `len(omi.history)`